Add base button block

yscik opened the pull request:
Provide a button block base that can be used for the upcoming button blocks.

### Changes proposed in this Pull Request

* Move block registration and entry files up a directory, as `blocks/index.js` and  `blocks/single-course.scss` for now.
* Add a Button builder function that contains an extendable base block definition
* Add a simple Take course button to use this
* Add the block radius code from Gutenberg core's button component
* Add support for for the new built-in color support key
  * This adds color settings, including gradients, and style rendering
  * We then remove the style rendering, because it applies the style to the block wrapper element :)
  * On older Wordpress versions, fall back to other color settings helpers we already use for the outline blocks. (No gradients)
* Add alignment support
  * Full width default is only for the Take course button, but it's just CSS
* Add fill and outline block styles (uses built-in styling)

### Testing instructions

* Create a course and add a Take Course button
* Customize it, save & publish/preview course
* Should look the same on the frontend
* Button should work and retain all settings after reloading the editor
* Button should mostly look and work the same as the core button across various themes

### TODO
- [ ] Unit test

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video

<img width="629" alt="image" src="https://user-images.githubusercontent.com/176949/95376502-78a9f480-08e1-11eb-8a6e-ee4761d2f2c0.png" align="center">

<img width="820" alt="image" src="https://user-images.githubusercontent.com/176949/95376763-ddfde580-08e1-11eb-9bea-75784b515cf8.png" align="center">

yscik commented on assets/blocks/button/color-hooks.js:
```diff
@@ -0,0 +1,124 @@
+export const addColorSettings = ( settings ) => {
+	if ( ! isSenseiButton( settings ) ) return settings;
+
+	if ( settings.supports?.color && ! settings.attributes?.backgroundColor ) {
+		settings.edit = withColorSettings( {
```
If the `supports: {color: true}` option is supported by current Gutenberg, it will have already added `backgroundColor` (and a few others here). If it's not there, wrap the edit component it a HOC that hooks up the color settings panel and attributes, and add the attributes needed. (Which are a bit different, having an extra `custom*` one for each color)

yscik commented on assets/blocks/button/color-hooks.js:
```diff
@@ -0,0 +1,124 @@
+ * @param {Object} props Block wrapper extra props.
+ * @return {Object} Block wrapper extra props.
+ */
+function removeColorProps( props ) {
+	return {
+		...props,
+		className: props.className?.replace(
+			/\s*has-[\w-]*(color|background)/g,
+			''
```
A tiny problem with the color support hook is it adds the color classes and styles to the wrapper component both in the editor and on save. Since we want those on our button only, we remove them here.

yscik commented on assets/blocks/button/color-props.js:
```diff
@@ -0,0 +1,105 @@
+ * @param {Object} props.attributes Block Attributes.
+ * @param {Object} props.colors     Theme color settings.
+ */
+export const getColorAndStyleProps = ( { attributes, colors } ) => {
```
This is the same code running in the color support hook that processes the selected color settings stored as attributes and builds CSS classes and inline styles. We use this to do the same mapping and use the styles for the inner button element.
(Core button copies this one too, for the same reasons.)

yscik commented on assets/blocks/button/index.js:
```diff
@@ -0,0 +1,75 @@
+ * @param {Object} opts
+ * @param {Object} opts.settings Block settings.
+ */
+export const buttonBlock = ( { settings, ...options } ) => {
+	options = {
+		tagName: 'a',
+		...options,
+	};
```
Settings are merged into block settings (like block name, title), the rest of options are passed to the edit and save functions. We can use this to make parts of the base component configurable/extendable. Right now it only has a `tagName` option for the output element.

yscik commented on assets/blocks/button/settings-button.js:
```diff
@@ -0,0 +1,67 @@
+ * @param {number?}  props.borderRadius  Border radius attribute.
+ * @param {Function} props.setAttributes Set block attributes.
+ */
+export const BorderPanel = ( { borderRadius, setAttributes } ) => {
+	return (
```
Small border radius control from core button block.

gkaragia reviewed (commented):
Looks great! Love where this is going! Couple of minor issues:
- The background doesn't respect the border radius when a solid color is used. It seems to work for gradient.
- The border radius has a default value of 5 although it looks like it has a 50.
- The class is displayed when Fill or Outline is selected in 'Additional CSS class(es)'

gkaragia commented on includes/blocks/class-sensei-course-outline-block.php:
```diff
@@ -44,10 +44,10 @@ public function enqueue_block_assets() {
 		}
 
-		Sensei()->assets->enqueue( 'sensei-course-outline', 'blocks/course-outline/style.css' );
+		Sensei()->assets->enqueue( 'sensei-single-course', 'blocks/single-course.css' );
```
Just saw the refactor in #3697, we could also do that there depending on what gets merged first.

yscik replied:
Yeah, good idea. I'll add a dedicated PR for it after both to keep it cleaner.

yscik commented on assets/blocks/button/color-hooks.js:
```diff
@@ -0,0 +1,124 @@
+ * @return {Object} props Extra save props.
+ */
+export function removeColorSaveProps( props, blockType ) {
+	if ( ! isSenseiButton( blockType ) ) return props;
```
Fixed in 0612778

yscik commented on assets/blocks/button/index.js:
```diff
@@ -0,0 +1,75 @@
+ * @param {Object} opts
+ * @param {Object} opts.settings Block settings.
+ */
+export const buttonBlock = ( { settings, ...options } ) => {
```
Updated to `createButtonBlockType` in d23c140

yscik commented on assets/blocks/course-outline/index.js:
```diff
@@ -1,10 +1,5 @@
-	const { name, ...settings } = block;
-	registerBlockType( name, settings );
-} );
+import '../button/index.js';
```
Cleaned that up in faff69d

yscik commented:
>  The background doesn't respect the border radius when a solid color is used. It seems to work for gradient.

Fixed this in 648dd9f, looks like there is another way to add wrapper styles 😨 
This is also testable with setting the alignment to left/right/center - only the now small button should have the colors.

>  The border radius has a default value of 5 although it looks like it has a 50.

Added a clearer 'unset' state in 3bcab8f. The default here is the theme's border radius, but we can't really figure out here what number that actually is. 
 
> The class is displayed when Fill or Outline is selected in 'Additional CSS class(es)'

Yeah, this is just how block styles work - it just adds the style to the classlist.

gkaragia reviewed (dismissed):
Looks great! There is a minor issue where when the block is inserted, the radius value isn't loaded in the input box. Happy to fix this as part of a later PR though.

yscik commented:
Just one more rebase to fix the merge conflict, and a small change to not open the border radius panel by default.

> the radius value isn't loaded in the input box.

The default border-radius here is what the theme sets in CSS. It's a bit tricky to get that value in JS — but even if we would do that, it's more clear to display this as empty when there is no value set in the control. (Similarly how background/text color is not displayed when nothing is set)

gkaragia reviewed (approved):
👌
